iquip/apps/dataviewer.py

```diff
@@ -634,24 +733,213 @@ def datasetName(self) -> str:
+    def setDataset(
+        self,
+        dataset: np.ndarray,
+        parameters: List[str],
+        units: List[Optional[str]],
+    ):
+        """Sets the dataset to show and updates the axis viewer.
+        
+        Args:
+            See SimpleScanDataPolicy.
+        """
+        self.policy = SimpleScanDataPolicy(dataset, parameters, units)
+        self.frame.sourceWidget.setParameters(parameters, units)
+
+    @pyqtSlot(tuple)
+    def setAxis(self, axis: Sequence[int]):
+        """Given the axis information, draws the main plot.
+        
+        Args:
+            axis: See updateMainPlot().
+        """
+        self.axis = axis
+        dataType = self.frame.dataPointWidget.dataType()
+        self.updateMainPlot(axis, dataType)
+
+    @pyqtSlot(DataPointWidget.DataType)
+    def setDataType(self, dataType: DataPointWidget.DataType):
+        """Given the data type, draws the main plot.
+        
+        Args:
+            dataType: See updateMainPlot().
+        """
+        self.updateMainPlot(self.axis, dataType)
+
+    def updateMainPlot(self, axis: Sequence[int], dataType: DataPointWidget.DataType):
+        """Updates the main plot.
+        
+        Args:
+            axis: See SimpleScanDataPolicy.extract().
+            dataType: Target data type.
+        """
+        if self.policy is None:
+            return
+        reduce = self._reduceFunction(dataType)
+        data, axes = self.policy.extract(axis, reduce)
+        self.frame.mainPlotWidget.setData(data, axes)
+        index = self.dataPointIndex
+        if data.ndim == len(index) and np.all(np.less(index, data.shape)):
+            self.selectDataPoint(index)
+        else:
+            self.selectDataPoint((0,) * data.ndim)
+
+    @pyqtSlot(tuple)
+    def selectDataPoint(self, index: Tuple[int, ...]):
+        """Selects a data point at the given index.
+        
+        Args:
+            index: The index of the target data point, in the dataset array.
+        """
+        if self.policy is None:
+            return
+        self.dataPointIndex = index
+        _, symbols = self.policy.symbolize(self.axis)
+        data_indices = np.all(symbols.T == index, axis=1)
+        data = self.policy.dataset[:, 0][data_indices].astype(int)
+        for dataType in DataPointWidget.DataType:
+            value = self._reduceFunction(dataType)(data)
+            self.frame.dataPointWidget.setValue(value, dataType)
+        bins, counts = np.unique(data, return_counts=True)
+        self.frame.dataPointWidget.setHistogramData(bins, counts)
+
+    @pyqtSlot(int)
+    def setThreshold(self, threshold: int):  # pylint: disable=unused-argument
+        """Given the threshold, updates the data point widget and the main plot.
+        
+        Args:
+            threshold: See DataPointWidget.setThreshold().
+        """
+        self.updateMainPlot(self.axis, self.frame.dataPointWidget.dataType())
+        self.selectDataPoint(self.dataPointIndex)
```

Comment:
Even if the threshold is changed, won't the data point plot be changed?

Reply:
Do you mean it shouldn't be changed? To be honest I was too lazy to update the P1 spinbox, so I just updated it all.
I improved the code to update only the P1 value!

Reply:
> Do you mean it shouldn't be changed?

Yes!